### PR TITLE
FIxes the multigpu setup thanks to bobby

### PR DESCRIPTION
**Pull request by DylanASHillier**
* Uses no_sync to speed stuff up (20%)
* Uses the appropriate random shuffling (80%)
* fixes an issue with ports and master address environment variables
* [currently] adds support to ignore all the distributed stuff for single gpu... should maybe be removed now that it works on both ...

### CODE REVIEW

File: trainers/build_trainers.py [hunk 1]
```diff
@@ -35,8 +35,13 @@ def ddp_setup(rank, world_size):
         rank: Unique identifier of each process
         world_size: Total number of processes
     """
-    os.environ["MASTER_ADDR"] = "localhost"
-    os.environ["MASTER_PORT"] = "12355"
+    # Get the master address and port from SLURM environment variables
+    master_addr = os.environ.get('MASTER_ADDR', 'localhost')
```
Comment (DylanASHillier) [inline trainers/build_trainers.py:39]: this allows us to use multi multigpu runs at a time

File: train.py [hunk 2]
```diff
@@ -49,6 +49,27 @@ def ddp_main(rank, world_size, cfg):
         # restore the print function
         restore_print_override(original_print)
 
+def single_gpu_main(cfg):
```
Comment (DylanASHillier) [inline train.py:52]: should keep this stuff or not?

**Comment by DylanASHillier** [comment]
[Troubleshoot load_data('openwebtext') as it's causing an OOM issue for the DDP process, let alone without DDP.](https://trello.com/c/SMh3Z8pJ/7-troubleshoot-loaddataopenwebtext-as-its-causing-an-oom-issue-for-the-ddp-process-let-alone-without-ddp)

File: configs/full_configs/baseline.yaml [hunk 3]
```diff
@@ -27,7 +27,7 @@ model:
   vocab_size: 50257
   model_shell_type: standard
   embedding_weight_tying: true
-  positional_encoding_type: rope
+  positional_encoding_type: learned
```
Comment (DylanASHillier) [inline configs/full_configs/baseline.yaml:30]: need to remove before merging

File: trainers/build_trainers.py (hunk 1 above)
Comment (bobbycxy) [reply trainers/build_trainers.py:39]: yea this is better. It's more dynamic

File: train.py (hunk 2 above)
Comment (bobbycxy) [reply train.py:52]: hmm, this seems duplicative and suggestive that our DDP isn't reliable. That's coz DDP should work even when GPU availability is 1. It'll work as a fail safe, but I think users will be doubtful if we've implemented other things reliably.

Comment (bobbycxy) [reply train.py:52]: So, I'm in favour not to keep it. Happy to hear @LeonGuertler 's thoughts

File: models/components/positional_encoding.py [hunk 4]
```diff
@@ -44,16 +45,44 @@ def forward(self, x):
         """
         return x
 
+class SinCosPosEncoding(
```
Comment (DylanASHillier) [inline models/components/positional_encoding.py:48]: lol this doesn't belong on this pr but shhh

Comment (bobbycxy) [reply models/components/positional_encoding.py:48]: yea.. I find it so too - might just be better to separate it lah

Comment (DylanASHillier) [reply models/components/positional_encoding.py:48]: aiya tai mafan

File: trainers/dataloader.py [hunk 5]
```diff
@@ -66,11 +68,12 @@ def __getitem__(self, idx):
         
         Note: This works for BytePooling and StandardDataloader, but not for ConversationalDataloader and NextTokenMLMDataloader
         '''
+        idx = random.randint(0, len(self.data) - self.context_window) # lol
```
Comment (bobbycxy) [inline trainers/dataloader.py:71]: think we can append a comment here that goes something like this? 

"## implementing our own random sampling here to avoid issues with Pytorch's DataLoader". So that no one accidentally thinks this is redundant and deletes it.

File: trainers/utils.py [hunk 6]
```diff
@@ -208,14 +208,23 @@ def forward_wrapper(*args, **kwargs):
...
     """
     Since using DDP, calculation of metrics happen across all GPUs. 
     This function aggregate the loss across all GPUs. 
     """
+    if not is_dist():
+        return value
     all_loss = torch.tensor([value], device=device)
     dist.all_reduce(all_loss, op=dist.ReduceOp.SUM)
     return all_loss.item() / dist.get_world_size()
+    # return value
```
Comment (bobbycxy) [inline trainers/utils.py:227]: prolly can remove this comment then?

File: trainers/base_trainer.py [hunk 7]
```diff
@@ -39,12 +40,17 @@ def __init__(
...
         dropout_scheduler=None,
     ) -> None:
         self.model = model
-        self.model = DDP(self.model, device_ids=[gpu_id])
+        if gpu_id is not None: # using ddp
+            self.dist = True
+            self.DDP_model = DDP(self.model, device_ids=[gpu_id])
+        else:
+            self.dist = False
+            self.DDP_model = model
```
Comment (bobbycxy) [inline trainers/base_trainer.py:53]: no need this alr right? since DDP should work with single GPU

**Review by bobbycxy — APPROVED** [review]
boss, looks good. Thank you for speeding it up!!
